### PR TITLE
Implementation of the join operation

### PR DESCRIPTION
**Pull request by dns2utf8**
it restricts the operation to observer only.

Because the `std::sync::mpsc::channel` does not offer a way to query the queue length I added another counter field: stored_jobs_counter

### CODE REVIEW

**Comment by frewsxcv** [comment]
What about if we use `Condvar` for this?

https://doc.rust-lang.org/std/sync/struct.Condvar.html

Seems like we could store a `Condvar` in the `ThreadPool`, and if the user calls `join`, we call `wait` on the `Condvar`, and every time the `ThreadPool` finishes, we call `notify_all`.

**Comment by dns2utf8** [comment]
Good point, we trade polling for a little memory. I am working on it.

**Comment by dns2utf8** [comment]
The implementation could be optimized further by moving more/all of the data to `ThreadPoolSharedData`.

While thinking about the problem I was thinking what if the pool were to be used to sync other threads like in the example below?
It currently does not work and it would require the `Sync` flag:
```rust
unsafe impl Sync for ThreadPool { }
```

```rust
        let pool = Arc::new(
                    ThreadPool::new_with_name("multi join test", 8));
        let test_count = Arc::new(AtomicUsize::new(0));

        let pool_r0 = pool.clone();
        let pool_r1 = pool.clone();
        let test_count_r0 = test_count.clone();
        let test_count_r1 = test_count.clone();

        let t0 = thread::spawn(move||{
            for _ in 0..21 {
                let test_count = test_count_r0.clone();
                pool_r0.execute(move || {
                    sleep(Duration::from_secs(2));
                    test_count.fetch_add(1, Ordering::Release);
                });
            }
            pool_r0.join();
            assert_eq!(42, test_count_r0.load(Ordering::Acquire));
        });

        let t1 = thread::spawn(move||{
            for _ in 0..21 {
                let test_count = test_count_r1.clone();
                pool_r1.execute(move || {
                    sleep(Duration::from_secs(2));
                    test_count.fetch_add(1, Ordering::Release);
                });
            }
            pool_r1.join();
            assert_eq!(42, test_count_r1.load(Ordering::Acquire));
        });

        
        pool.join();
        assert_eq!(42, test_count.load(Ordering::Acquire));

        t0.join().unwrap();
        t1.join().unwrap();
```

**Comment by frewsxcv** [comment]
> The implementation could be optimized further by moving more/all of the data to `ThreadPoolSharedData`.

Is it possible you could pull this change out into a separate pull request? We can merge that first _without_ this `join` method. Then we can rebase this `join` pull request off those changes once those merge.

> Open interface new_with_name to accept anything convertable into String

Regarding this commit, see my response in https://github.com/frewsxcv/rust-threadpool/pull/58

**Comment by dns2utf8** [comment]
Well it is going to take some time, since I decided to create the shared data struct when I realized the many times I would have to pass around 6 counters for every operation and handle their inconsistent names after I implemented the `join`. Let me see what I can do.

**Comment by dns2utf8** [comment]
I created two new branches the PRs with one commit each. It was simpler than doing some git magic 😉

**Comment by frewsxcv** [comment]
[This usage of `sleep` can be replaced with `join`, right?](https://github.com/dns2utf8/rust-threadpool/blob/cc5fc936e71954c1e938dd8d6bc872c22bf5d6ac/lib.rs#L295). If so, there are a few other examples and tests that can be replaced:

https://github.com/dns2utf8/rust-threadpool/blob/cc5fc936e71954c1e938dd8d6bc872c22bf5d6ac/lib.rs#L457
https://github.com/dns2utf8/rust-threadpool/blob/cc5fc936e71954c1e938dd8d6bc872c22bf5d6ac/lib.rs#L478
https://github.com/dns2utf8/rust-threadpool/blob/cc5fc936e71954c1e938dd8d6bc872c22bf5d6ac/lib.rs#L492
https://github.com/dns2utf8/rust-threadpool/blob/cc5fc936e71954c1e938dd8d6bc872c22bf5d6ac/lib.rs#L691

**Comment by dns2utf8** [comment]
Yes, let me replace them quickly. As mentioned by @bblancha this should solve #66 as well.

**Comment by dns2utf8** [comment]
I added support for a pool where all threads panic. This would have lead to a stall until at least one job would have succeeded.

Some of the `sleep`s you mentioned are measuring the `active_count`. Joining the pool would cause the number to be zero so I left them as they were. 

Do you see any unhandled cases that we should test for?

File: lib.rs [hunk 1]
```diff
@@ -137,11 +138,29 @@ impl<'a> Drop for Sentinel<'a> {
...
 
+impl ThreadPoolSharedData {
+    fn has_work(&self) -> bool {
+        self.stored_jobs_counter.load(Ordering::SeqCst) > 0
+            || self.active_count.load(Ordering::SeqCst) > 0
+    }
+
+    /// Notify all observers joining this pool if there is no more work to do.
+    fn no_work_notify_all(&self) {
+        if self.has_work() == false {
```
Comment (frewsxcv) [inline lib.rs]:
```rust
if !self.has_work()
```

File: lib.rs [hunk 2]
```diff
@@ -310,6 +333,43 @@ impl ThreadPool {
             }
         }
     }
+    
+    /// Block the current thread until all jobs in the pool are completed.
+    /// Once waiting for the pool to complete you can no longer add new jobs, &mut self ensures that.
+    ///
+    /// ```
+    /// # use threadpool::ThreadPool;
```
Comment (frewsxcv) [inline lib.rs]: I think we should uncomment this so we have a complete example a user can just copy/paste

File: lib.rs [hunk 3]
```diff
@@ -128,6 +128,7 @@ impl<'a> Drop for Sentinel<'a> {
             self.shared_data.active_count.fetch_sub(1, Ordering::SeqCst);
             if panicking() {
                 self.shared_data.panic_count.fetch_add(1, Ordering::SeqCst);
+                self.shared_data.no_work_notify_all();
```
Comment (frewsxcv) [inline lib.rs]: If this line was moved _outside_ this `if panicking()` conditional, could we get rid of [this line](https://github.com/frewsxcv/rust-threadpool/pull/63/files#diff-8c13b5c367b1a0d3dc04da468ea844bdR426)? Would that be the same thing?

Comment (dns2utf8) [inline lib.rs]: No, because a worker thread keeps on working on jobs for as long as possible and reducing the expensive create/cleanup cycle of threads.

File: lib.rs [hunk 4]
```diff
@@ -207,6 +226,9 @@ impl ThreadPool {
         let rx = Arc::new(Mutex::new(rx));
 
         let shared_data = Arc::new(ThreadPoolSharedData {
+            empty_condvar: Condvar::new(),
+            empty_trigger: Mutex::new(false),
+            stored_jobs_counter: AtomicUsize::new(0),
```
Comment (dns2utf8) [inline lib.rs]: sure

File: lib.rs (hunk 2 above)
Comment (dns2utf8) [inline lib.rs]: 👍

File: lib.rs [hunk 5]
```diff
@@ -310,6 +333,43 @@ impl ThreadPool {
             }
         }
     }
+    
+    /// Block the current thread until all jobs in the pool are completed.
+    /// Once waiting for the pool to complete you can no longer add new jobs, &mut self ensures that.
```
Comment (dns2utf8) [inline lib.rs]: The reason I did it in the first place was because without the `mut` many threads can `join` at the same time.

Since the `join` is blocking, the current thread can not send new jobs anyways.
Accessing the pool from many threads requires the `Sync` flag. But I have to reason about that some more because there is an edge case, [see this test](https://github.com/frewsxcv/rust-threadpool/pull/63/commits/78e54dd79ced4613e6fcaa526ff08d1ef6846d18#diff-8c13b5c367b1a0d3dc04da468ea844bdR737).

**Comment by dns2utf8** [comment]
I added test for the new cases. The PR is currently mergable up to 254fa9da137f00b.

**Comment by dns2utf8** [comment]
I found a very nasty deadlock with the `fn test_multi_join()`. The reason was some threads could end up in an infinite loop because they did not exit the false-positive loop fast enough. Since testing for the `Condvar` requires the thread to acquire the `Mutex` `shared_data.empty_trigger` they would block until the thread was joinable.

While debugging I added the `.field("queued_count", &self.queued_count())` to the debug output and exposed the `queued_count` like the other counters with a function.

What do you think?

**Comment by frewsxcv** [comment]
LGTM, thanks for doing this!

**Comment by dns2utf8** [comment]
It is a pleasure solving the hard problems 😄
